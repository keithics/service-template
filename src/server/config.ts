/**
 * CONFIG file, either loaded via .env file or system ENV
 */
const { DATABASE_URL, LOGTAIL_TOKEN, MONGO_DEBUG, NODE_HOST, NODE_PORT, ADMIN_JWT, TOKEN_SECRET_ADMIN, TOKEN_SECRET, SENTRY } = process.env;

const config = {
  MONGO_URL: DATABASE_URL || 'mongodb://localhost:27017/signalytics',
  MONGO_DEBUG: MONGO_DEBUG || false,
  LOGTAIL_TOKEN: LOGTAIL_TOKEN || 'xxxxx',
  PROJECT_NAME: 'NOTIFICATION SERVICE',
  NODE_HOST: NODE_HOST || 'localhost',
  NODE_PORT: NODE_PORT || 8080,
  // we use this if we run the services locally and in order not to have a conflict with other service's port number
  NODE_PORT_ALT: 3000,
  TOKEN_SECRET,
  TOKEN_SECRET_ADMIN,
  ADMIN_JWT,
  SENTRY,
};

export default config;
