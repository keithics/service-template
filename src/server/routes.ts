import { Router } from 'express';
import config from './config';
import pokemonRouter from 'api/pokemons/pokemon.routes';
import { authenticate } from '@signalyticsai/auth/lib/authenticate';

const routes = Router();

routes.use('/', authenticate({ token: config.TOKEN_SECRET, level: 'user', roles: ['my-custom-role'] }), pokemonRouter);

// 404 error
routes.use('*', (req, res) => {
  res.status(404).json({ error: true, message: '404 not found' });
});

export default routes;
