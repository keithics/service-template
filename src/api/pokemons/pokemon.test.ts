import request from 'supertest';
import { mongo } from 'server/mongo';
import app from 'server/app';
import createFakeUser from '@keithics/core/lib/jest.fake-user';
import jsonUpdate from './json/pokemon-update.json';
import json from './json/pokemon.json';

let token = '';
let createdId = '';
jest.mock('libraries/helpers');
jest.mock('server/config');

beforeAll(async () => {
  await mongo.dropAllCollections();
  // create user
  const fakeUser = await createFakeUser(false);
  token = fakeUser.token;
});

afterAll((done) => {
  (async () => {
    await mongo.dropAllCollections();
    await mongo.close();
  })();
  done();
});

describe('Pokemon Tests', () => {
  describe('CRUD routes', () => {
    test('CREATE - Should respond with status code 200', async () => {
      const response = await request(app)
        .post('/')
        .send(json)
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
      expect(response.body).toMatchSnapshot();
      createdId = response.body._id;
    });

    test('CREATE WITHOUT TOKEN - Should respond with status code 401', async () => {
      const response = await request(app).post('/').send(json);
      expect(response.status).toEqual(401);
    });

    test('READ ONE  - Should respond with status code 200', async () => {
      const response = await request(app)
        .get(`/${createdId}`)
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
      expect(response.body).toMatchSnapshot();
    });

    test('READ ONE WITHOUT TOKEN - Should respond with status code 401', async () => {
      const response = await request(app).get(`/${createdId}`);
      expect(response.status).toEqual(401);
    });

    test('PAGINATION - Should respond with status code 401', async () => {
      const response = await request(app)
        .post('/page')
        .send({ page: 1 })
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
      expect(response.body).toMatchSnapshot();
    });

    test('PAGINATION WITHOUT TOKEN - Should respond with status code 401', async () => {
      const response = await request(app).get('/page/1');
      expect(response.status).toEqual(401);
    });

    test('FILTERING  - Should respond with status code 200', async () => {
      const response = await request(app)
        .post(`/filter`)
        .send({
          filters: [
            {
              key: 'type',
              value: 'electric',
            },
          ],
        })
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
      expect(response.body).toMatchSnapshot();
    });

    test('SEARCHING  - Should respond with status code 200', async () => {
      const response = await request(app)
        .post(`/search`)
        .send({
          key: 'name',
          value: 'pika',
        })
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
      expect(response.body).toMatchSnapshot();
    });

    test('UPDATE ONE  - Should respond with status code 200', async () => {
      const response = await request(app)
        .put(`/${createdId}`)
        .send(jsonUpdate)
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
      expect(response.body).toMatchSnapshot();
    });

    test('UPDATE ONE WITHOUT TOKEN  - Should respond with status code 401', async () => {
      const response = await request(app).put(`/${createdId}`).send(jsonUpdate);
      expect(response.status).toEqual(401);
    });

    test('DELETE ONE  - Should respond with status code 200', async () => {
      const response = await request(app)
        .delete(`/${createdId}`)
        //
        .set('Authorization', `Bearer ${token}`);
      expect(response.status).toEqual(200);
    });

    test('DELETE ONE WITHOUT TOKEN - Should respond with status code 401', async () => {
      const response = await request(app).delete(`/${createdId}`);
      expect(response.status).toEqual(401);
    });
  });

  // end of v1
});
