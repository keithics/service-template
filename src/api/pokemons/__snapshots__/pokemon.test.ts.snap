// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Tests CRUD routes CREATE - Should respond with status code 200 1`] = `
Object {
  "__v": 0,
  "_id": "000000000000000000000000",
  "createdAt": "1984-01-24T16:00:00.000Z",
  "name": "Pikachu",
  "type": "electric",
  "updatedAt": "1984-01-24T16:00:00.000Z",
  "user": "000000000000000000000000",
}
`;

exports[`Pokemon Tests CRUD routes FILTERING  - Should respond with status code 200 1`] = `
Object {
  "docs": Array [
    Object {
      "__v": 0,
      "_id": "000000000000000000000000",
      "createdAt": "1984-01-24T16:00:00.000Z",
      "name": "Pikachu",
      "type": "electric",
      "updatedAt": "1984-01-24T16:00:00.000Z",
      "user": "000000000000000000000000",
    },
  ],
  "hasNextPage": false,
  "hasPrevPage": false,
  "limit": 15,
  "nextPage": null,
  "page": 1,
  "pagingCounter": 1,
  "prevPage": null,
  "totalDocs": 1,
  "totalPages": 1,
}
`;

exports[`Pokemon Tests CRUD routes PAGINATION - Should respond with status code 401 1`] = `
Object {
  "docs": Array [
    Object {
      "__v": 0,
      "_id": "000000000000000000000000",
      "createdAt": "1984-01-24T16:00:00.000Z",
      "name": "Pikachu",
      "type": "electric",
      "updatedAt": "1984-01-24T16:00:00.000Z",
      "user": "000000000000000000000000",
    },
  ],
  "hasNextPage": false,
  "hasPrevPage": false,
  "limit": 15,
  "nextPage": null,
  "page": 1,
  "pagingCounter": 1,
  "prevPage": null,
  "totalDocs": 1,
  "totalPages": 1,
}
`;

exports[`Pokemon Tests CRUD routes READ ONE  - Should respond with status code 200 1`] = `
Object {
  "_id": "000000000000000000000000",
  "createdAt": "1984-01-24T16:00:00.000Z",
  "name": "Pikachu",
  "type": "electric",
  "updatedAt": "1984-01-24T16:00:00.000Z",
  "user": "000000000000000000000000",
}
`;

exports[`Pokemon Tests CRUD routes SEARCHING  - Should respond with status code 200 1`] = `Array []`;

exports[`Pokemon Tests CRUD routes UPDATE ONE  - Should respond with status code 200 1`] = `
Object {
  "_id": "000000000000000000000000",
  "createdAt": "1984-01-24T16:00:00.000Z",
  "name": "mimikyu",
  "type": "dark",
  "updatedAt": "1984-01-24T16:00:00.000Z",
  "user": "000000000000000000000000",
}
`;
